Add Terraform deployment engine for cloud resources

jimright opened the pull request:
Added option for Terraform deployment of the cloud resources at the infrastructure and platform run-levels.
Features:
* A parameter `infra_deployment_engine` used to determine the choice of deployment engine (pure-Ansible or Terraform) for cloud resources.
* Terraform deployment is driven by Ansible, which creates necessary .tf resource files from Jinja templates and runs terraform apply\destroy. 
* Currently only supports the AWS `infra_type`:
  * Creates network, storage and compute resources at the infra level
  * Creates the AWS IAM policies and roles at the plat level
* Support for `local` and `remote_s3` (S3 bucket for storage; DynamoDB for state locking) backend storage of the Terraform state.
* A timestamped artefact directory with a copy of the Terraform files created at each execution of cloudera-deploy.

Draft PR has a `debug_terraform` bool variable. This puts pause points before the main terraform steps allowing for review. These will be cleaned before the final PR.

jimright commented:
Have rebased the code in this PR following the 1.5.1 release and retested that L0 and L1 deployment and teardown with Terraform work ok.

Chaffelson commented on roles/platform/tasks/initialize_setup_aws.yml:
```diff
@@ -78,6 +78,7 @@
 - name: Handle AWS Public and Private VPC Subnets if not defined
   when: not plat__aws_public_subnet_ids or not plat__aws_private_subnet_ids
   block:
+    # Search using a wildcard on the subnet Name tag {{ plat__namespace }}*
```
I'm confused by this comment

jimright replied:
Sorry, a hold-over from an older version of my code. Removed the comment now.

Chaffelson commented on roles/platform/tasks/setup_aws_terraform_authz.yml:
```diff
@@ -0,0 +1,84 @@
+  when: debug_terraform | default(false) | bool
+
+- name: Applying Terraform
+  community.general.terraform:
```
logging?

Chaffelson commented on roles/platform/tasks/teardown_aws_terraform_authz.yml:
```diff
@@ -0,0 +1,78 @@
+
+- name: Destroy Terraform authz resources
+  community.general.terraform:
+    project_path: "{{ plat__terraform_workspace_dir }}/workspace/plat"
```
logz

Chaffelson commented on roles/platform/tasks/teardown_aws_terraform_authz.yml:
```diff
@@ -0,0 +1,78 @@
+    path: "{{ plat__terraform_workspace_dir }}/workspace/plat"
+    state: absent
+    
+# TODO: Figure out how to handle below conditional with Terraform
```
TBD?

jimright replied:
Yes I need to figure out how to skip teardown of Terraform-created resources if some flags (e.g. env.teardown.delete_cross_account: False) are specified. I've removed the TODO from the code and will create an issue for these tasks.

jimright commented on roles/common/defaults/main.yml:
```diff
@@ -57,6 +57,7 @@ common__ngw_suffix:                       "{{ globals.labels.nat_gateway | defau
 common__unique_storage_name_suffix:       "{{ globals.storage.name | default((common__region + common__aws_profile) if 'aws' in common__infra_type else common__region) }}"
 
 # Infra
+common__infra_deployment_engine:          "{{ globals.infra_deployment_engine | default() }}"
```
Good spot, changed that in the latest update of the code.

jimright commented on roles/common/defaults/main.yml:
```diff
@@ -66,6 +67,19 @@ common__public_key_text:                  "{{ globals.ssh.public_key_text | defa
 
+# Terraform
+# The processed Jinja template files for Terraform are placed in common__terraform_template_dir
+common__terraform_template_dir:            "{{ globals.terraform_template_dir | default( ( playbook_dir , 'terraform_processed_template_code') | path_join ) }}"
```
Great idea. I've added the `terraform_base_dir` variable and all other Terraform directories are derived from this.

jimright commented on roles/infrastructure/tasks/setup_terraform.yml:
```diff
@@ -0,0 +1,84 @@
+    prompt: "Ready to do Terraform apply on infra files in {{ infra__terraform_workspace_dir }}/workspace/infra"
+  when: debug_terraform | default(false) | bool
+
+- name: Applying Terraform
```
I'm not sure about how to increase the verbosity by default. Will have to look into this a bit more.

jimright commented on roles/infrastructure/tasks/setup_terraform.yml:
```diff
@@ -0,0 +1,84 @@
+    state: absent
+  when: infra__terraform_state_storage in ['remote_s3']
+
+# If created Utility Instance via Terraform then
```
The Terraform and Ansible set of tasks are a bit different for this. Ansible creates the utility instance and adds it to the inventory; Terraform discovers the utility instance and adds to inventory. So I think it best to keep these tasks separate.

jimright commented on roles/infrastructure/tasks/validate_aws_terraform.yml:
```diff
@@ -0,0 +1,56 @@
+  - name: Check infra__terraform_template_dir
+    ansible.builtin.assert:
+      that: 
+        - "infra__terraform_template_dir is defined"
```
Good spot, changed that in the latest update of the code.

jimright commented on roles/platform/tasks/initialize_aws_terraform.yml:
```diff
@@ -0,0 +1,94 @@
+    state: directory
+  register: __aws_policy_tmpdir
+
+- name: Download AWS default policy documents
```
Agree. In my updated changes, I've moved the download and process of AWS policy documents to a single task file, `aws_policy_download.yml`. 
I changed the below task files to then include this task file:
* setup_aws_autz.yml
* initialize_aws_terraform.yml, 
* initialize_teardown_aws_terraform.yml

jimright commented:
Hey @Chaffelson 

Thanks for looking through this. 
I've changed and updated cloudera.exe based on your comments. 
See above replies to your comments for specific details of changes.

Together with the updates to cloudera-deploy, this should be ready to retest now.

jimright commented:
@Chaffelson @wmudge,

Have updated the cloudera.exe Terraform PR following the release of the AWS provider. 
This new provider version had some breaking changes and deprecations - listed [here](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/guides/version-4-upgrade) - so I've handled those and tested to confirm everything still works.

Thanks,
Jim

wmudge commented on roles/infrastructure/tasks/setup_terraform.yml:
```diff
@@ -0,0 +1,93 @@
+    state: absent
+  when: infra__terraform_state_storage in ['remote_s3']
+
+# Get information about Dynamic Inventory VMs if created via Terraform
```
Could this have been done using output variables from TF?

jimright replied:
Yes, I will explore if using tf output is possible and submit a separate PR for this.

wmudge commented on roles/infrastructure/tasks/setup_terraform.yml:
```diff
@@ -0,0 +1,93 @@
+  when: infra__create_utility_service
+  block:
+    - name: Discover the Utility Instance details
+      community.aws.ec2_instance_info:
```
Use output variables in TF?

jimright replied:
As above, I will test this and submit a separate PR if it looks good.

wmudge commented on roles/infrastructure/tasks/teardown_terraform.yml:
```diff
@@ -0,0 +1,45 @@
+- name: Check if Terraform workspace directory exists
+  ansible.builtin.stat: 
+    path: "{{ infra__terraform_workspace_dir }}/infra"
+  register: workdir
```
Where is this registered variable used?

jimright replied:
Good spot, this is no longer used. There are a few similar instances of this in the code that I have cleaned as part of 
287c30d.

wmudge commented on roles/infrastructure/tasks/main.yml:
```diff
@@ -15,5 +15,5 @@
 - include_tasks: validate.yml
-- include_tasks: initialize_setup.yml
-- include_tasks: setup.yml
+# - include_tasks: initialize_setup.yml
```
Why is this commented out?

jimright replied:
Apologies, this is a hangover from when I first started this PR and trying to understand cloudera.exe. Uncommented these lines in [617d68c](https://github.com/cloudera-labs/cloudera.exe/pull/56/commits/617d68c30954838b9e3c921de9e84d92dc594445)

jimright commented:
@wmudge. 

Thanks for looking through the PR. I have updated the code to address most of the issues you spotted. Also retested the AWS Terraform deployment and teardown to make sure that still works.

For things like using terraform output instead of Ansible-based discovery, I will test this separately and create a PR if all goes well.

Jim
